File: src/slidingwindow/MinimumSizeSubarraySum.java
package slidingwindow;


public class MinimumSizeSubarraySum {
    public static void main(String[] args) {
        MinimumSizeSubarraySum sum = new MinimumSizeSubarraySum();
        System.out.println("args = " + sum.minSubArrayLen(11,new int[]{1,1,1,1,1,1,1,1}));
    }

    public int minSubArrayLen(int target, int[] nums) {
        int start = 0;
        int end = -1;
        int windowSum = 0;
        int ans = Integer.MAX_VALUE;

        for (int i = 0; i < nums.length; i++) {
            windowSum += nums[i];
            end = i;
            if (windowSum + nums[i] >= target){
                //shrink
                while(windowSum >= target){
                    ans = Math.min(end-start+1,ans);
                    windowSum-= nums[start];
                    start++;
                }
            }
        }
        if (ans == Integer.MAX_VALUE) {
            return 0;
        }
        return ans;
    }
}

File: src/gojek/slidingwindow/MinimumSizeSubarraySum.kt
package gojek.slidingwindow

import kotlin.math.min

class MinimumSizeSubarraySum {
    fun minSubArrayLen(target: Int, nums: IntArray): Int {
        var sum = 0
        var ans = Integer.MAX_VALUE

        var start = 0

        nums.forEachIndexed{ index, value ->
            sum+= value

            while (sum >= target){
                ans = min(ans,index-start+1)
                sum -= nums[start]
                start++
            }
        }
        return if (ans == Integer.MAX_VALUE) 0 else ans
    }
}

fun main() {
    val a = MinimumSizeSubarraySum()
    val aa = intArrayOf(1,1,1,1,1,1,1,1)

    print(a.minSubArrayLen(11, aa))
}